ext/odbc/php_odbc.c

```diff
@@ -2246,7 +2246,17 @@ void odbc_do_connect(INTERNAL_FUNCTION_PARAMETERS, int persistent)
 				RETCODE ret;
 				UCHAR d_name[32];
 				SQLSMALLINT len;
+				SQLUINTEGER dead = SQL_CD_FALSE;
 
+				ret = SQLGetConnectAttr(db_conn->hdbc,
+					SQL_ATTR_CONNECTION_DEAD,
+					&dead, 0, NULL);
+				if (dead == SQL_CD_TRUE) {
+					/* Bail early here, since we know it's gone */
+					zend_hash_str_del(&EG(persistent_list), hashed_details, hashed_len);
+					goto try_and_get_another_connection;
+				}
+				/* If the driver doesn't support it, fall back to old heuristic */
```

Comment:
Hmm, can't we assume that the driver supports `SQL_ATTR_CONNECTION_DEAD` if `ret == SQL_SUCCESS`? I mean, can't we skip the fallback (`SQL_DATA_SOURCE_READ_ONLY`) regardless of the value of `dead`?

Reply:
Unfortunately, the Db2i driver at least had a case where it would return `SQL_CD_FALSE`, but the heuristic would detect a failed connection. However, that case (specifically, admin killing the database server process) is pretty weird and uncommon (though people bug me enough about it I have to mention it). You're probably right; it'd be better to just trust `SQL_ATTR_CONNECTION_DEAD`, and get driver authors to implement more accurate results if they're going to use it.